File: Java/src/main/java/Lists/NestedListWeightSum/NestedListWeightSum.java
package Lists.NestedListWeightSum;

import java.util.List;

public class NestedListWeightSum {

    public static void main(String [] args) {

    }

    public static int depthSum(List<NestedInteger> nestedList) {
        int level = 1;
        int total = traverseList(nestedList, level);

        return total;
    }

    private static int traverseList(List<NestedInteger> nestedList, int level) {

        int total = 0;
        for (NestedInteger element : nestedList){

            // Check if Integer
            if (element.isInteger()) {
                total += element.getInteger() * level;
                System.out.println("Integer Encountered: " + element.getInteger() + "       Total: " + total + "    Level: " + level);
                continue;
            }

            // If List
            total += traverseList(element.getList(), level + 1);
            System.out.println("List Encountered: " + element.getList() + "       Total: " + total + "    Level: " + level);
        }

        return total;
    }
}


/**
 Notes
 - A bit tricky but not tooo much
 - Uses recursion to traverse the inner Trees

 Solution
 1. Traverse the List
 2. Check if object is integer or list
 3. If list, use recursion to find the total
    > If integer, Add to the total
 4. Continue until all elements are traversed



 */

File: Python/project/src/Lists/NestedListWeightSum/NestedListWeightSum.py
import NestedInteger from NestedInteger.py      # This import is NOT required since the class itself is undefined



class Solution(object):
    def depthSum(self, nestedList):

        level = 1
        total = self.traverseList(nestedList, level)

        return total

    def traverseList(self, nestedList, level):
        total = 0

        for element in nestedList:

            if element.isInteger():
                total += element.getInteger() * level
                continue

            total += self.traverseList(element.getList(), level + 1)

        return total


def main():
    solution = Solution()

    nestedList = [[1,1],2,[1,1]]
    output = solution.depthSum(nestedList)
    print ("Weight Sum: ", output, "\n")

    nestedList = [1,[4,[6]]]
    output = solution.depthSum(nestedList)
    print ("Weight Sum: ", output)


if __name__ == '__main__':
    main()


'''
Notes
- Abit tricky but not too much. Had to solve in java first
- Uses recursion to traverse down the list

Solution
1. Traverse the list
2. If element is integer, add to total
3. If list, then we use recursion to find the total of its element
4. Continue until all elements have been iterated


'''